engine/robot.py

```diff
@@ -330,8 +374,97 @@ def set_phase(self, new_phase):
+                    self.avoid_obstacle = True
+                    with open(ROOT_DIR + '/tests/functionality_tests/csv/avoid_obstacle_result.csv', 'a') as fd:
+                        fd.write("Avoid" + '\n')
+                else:
+                    self.avoid_obstacle = False
+                    with open(ROOT_DIR + '/tests/functionality_tests/csv/avoid_obstacle_result.csv', 'a') as fd:
+                        fd.write("Not Avoid" + '\n')
+            if curr_ultrasonic_value < 0:
+                self.set_phase(Phase.FAULT)  # value should not go below 0; sensor is broken
+                with open(ROOT_DIR + '/tests/functionality_tests/csv/avoid_obstacle_result.csv', 'a') as fd:
+                    fd.write("Fault" + '\n')
+
+        # time.sleep(10)  # don't hog the cpu
+
+    def execute_avoid_obstacle(self, dist_to_goal, prev_phase, init_threshold, goal_threshold):
+        """ Execute obstacle avoidance
+            Args:
+                dist_to_goal (Double): The distance from the robot to the goal at the start of the phase
+                prev_phase (Phase): The previous phase to return to after executing obstacle avoidance
+                init_threshold (Double): Radius from initial position that robot has to leave before it can detect if
+                    goal is unreachable
+                goal_threshold (Double): Threshold from goal that will be detected as reaching goal
+            Returns:
+                prev_phase (Phase)
+        """
+        init_x = self.x_pos
+        init_y = self.y_pos
+        # threshold is arbitrary, set later. sometimes location reading will be inaccurate or not frequent enough to
+        # read that we've arrived back to the init x y pos. make it small enough that robot actually leaves threshold
+        # at some time during boundary following or add timeout to branch making gate = True
+        gate = False
+        dist_to_goal_decreases = False
+        boundary_traversed = False
+        while True:
+            if not self.fault:  # fault has priority over obstacle avoidance
+                dist_from_init = math.sqrt((self.x_pos - init_x) ** 2 + (self.y_pos - init_y) ** 2)
+                if dist_from_init > init_threshold:
+                    gate = True
+                if curr_dist_to_goal < goal_threshold and gate:  # exits obstacle avoidance if robot close to goal
+                    self.set_phase(prev_phase)
+                    return None
+                elif boundary_traversed and gate:
+                    self.set_phase(Phase.FAULT)  # cannot reach goal
+                    return None
+                elif dist_to_goal_decreases:
+                    # don't include condition on whether there is a new obstacle bc it will be caught and
+                    # current algorithm will continue traversing current boundary (instead of traversing new obstacle)
+                    # not optimal because frequently will have to go back to obstacle avoidance
```

Comment:
So we want to transition back to our previous phase as long as we are closer to the goal than we were previously?

I feel like this might lead to a lot of phase transitions in and out of the obstacle avoidance phase if the obstacle is still in the detection range of the robot. This could potentially lead to concurrency issues. If my understanding is correct, we might have the following scenario?

For example, consider the case where R = robot, G = Goal, O = Obstacle, and the goal is within the detection range and each | | is a cell/node after a timestep of moving forward.
|G|
|   |
|O|
|   |
 R
If the robot proceeds forward one time step, it hasn't successfully avoided the obstacle but its closer to the goal. So, we would switch to the previous phase (lets say traverse for example).
|G|
|   |
|O|
|R|
There could be a case where move_to_target_node begins executes before track_obstacle "interrupts" the traversal phase. Then collision would happen?
Or do we assume `dist_to_goal_decreases` being true means boundary following takes the robot out of the way from the obstacle first?

Reply:
we don't assume that dist_to_goal_decreases being true means boundary following takes the robot out of the way of the obstacle
To fix the concurrency issue, we can have another function where the robot turn towards the target and then check and then potentially obstacle avoiding again?

Reply:
Where would this check be?

Reply:
We can put this in a separate function that will be called (within the avoid obstacle algorithm) before returning to the traversal phase

Reply:
YUS